### ISSUE TITLE
影像向量化模組（雙向量輸出）

### ISSUE THREAD

**Issue by B30780** [B30780/AOI_Defect_LLM_CCPM#14]
name: 敶勗????芋蝯????撓?綽?
status: open
created: 2026-01-20T10:40:44Z
updated: 2026-01-20T10:40:44Z
github: [Will be updated when synced to GitHub]
depends_on: [001, 002]
parallel: true
conflicts_with: [005]

# Task: 敶勗????芋蝯????撓?綽?

## Description

撖虫??箸 CLIP ViT-B/32 ?蔣????璅∠?嚗?湧???頛詨璈嚗?憪蔣????+ ?桃蔗敶勗???嚗??Qdrant ??鞈?摨怠恥?嗥垢嚗祕?暸???????甈∟??迨璅∠??箸??AOI 蝻粹瑼Ｙ揣蝟餌絞?敹敺菜??惜??

**?詨??嚗?*
- CLIP ViT-B/32 璅∪??游????
- ??敶勗??蝵拙蔣????????
- Qdrant ??鞈?摨恍???雿?
- ??甇????璅?????
- ?寞活???????
- ??蝬剖漲嚗?12-dim

## Acceptance Criteria

- [ ] ???游? CLIP ViT-B/32 璅∪?嚗蝙??OpenAI CLIP ??transformers嚗?
- [ ] 撖衣??????original_vector (512-dim) ??mask_vector (512-dim)
- [ ] Qdrant 摰Ｘ蝡舀迤撣賊?銝血?瑁? CRUD ??
- [ ] ??甇????桐???嚗2 norm = 1嚗?
- [ ] ?桀撐敶勗???????< 5 蝘?
- [ ] ?寞活????湛?batch size >= 8嚗?
- [ ] ?????脣???Qdrant ??嚗??怠???metadata
- [ ] ?桀?皜祈岫閬???>= 80%
- [ ] ?航炊??嚗芋???亙仃?蔣?撘隤扎drant ??憭望?

### ?嗆?閮剛?

```python
# src/vectorization/clip_vectorizer.py
import torch
import clip
from PIL import Image
from typing import Tuple, List, Optional
import numpy as np
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct, VectorParams, Distance

class CLIPVectorizer:
    "
    CLIP-based image vectorizer with dual vector output support.
    Generates embeddings for both original and masked defect images.
    "
    
    def __init__(self, model_name: str = ViT-B/32, device: str = cuda):
        "
        Initialize CLIP model and preprocessing.
        
        Args:
            model_name: CLIP model variant (default: ViT-B/32)
            device: Computation device (cuda/cpu)
        "
        self.device = device
        self.model, self.preprocess = clip.load(model_name, device=device)
        self.model.eval()  # Set to evaluation mode
        
    def vectorize_image(self, image: Image.Image) -> np.ndarray:
        "
        Generate 512-dim embedding for a single image.
        
        Args:
            image: PIL Image object
            
        Returns:
            Normalized 512-dim numpy array
        "
        with torch.no_grad():
            image_input = self.preprocess(image).unsqueeze(0).to(self.device)
            embedding = self.model.encode_image(image_input)
            embedding = embedding / embedding.norm(dim=-1, keepdim=True)  # L2 normalize
            return embedding.cpu().numpy().flatten()
    
    def vectorize_dual(self, 
                       original_image: Image.Image, 
                       mask_image: Image.Image) -> Tuple[np.ndarray, np.ndarray]:
        "
        Generate dual vectors for original and masked images.
        
        Args:
            original_image: Original defect image
            mask_image: Masked defect region image
            
        Returns:
            Tuple of (original_vector, mask_vector), both 512-dim
        "
        original_vector = self.vectorize_image(original_image)
        mask_vector = self.vectorize_image(mask_image)
        return original_vector, mask_vector
    
    def batch_vectorize(self, images: List[Image.Image], 
                       batch_size: int = 8) -> np.ndarray:
        "
        Batch vectorization for efficiency.
        
        Args:
            images: List of PIL Image objects
            batch_size: Batch processing size
            
        Returns:
            Array of shape (N, 512)
        "
        embeddings = []
        for i in range(0, len(images), batch_size):
            batch = images[i:i+batch_size]
            with torch.no_grad():
                batch_input = torch.stack([
                    self.preprocess(img) for img in batch
                ]).to(self.device)
                batch_embeddings = self.model.encode_image(batch_input)
                batch_embeddings = batch_embeddings / batch_embeddings.norm(
                    dim=-1, keepdim=True
                )
                embeddings.append(batch_embeddings.cpu().numpy())
        
        return np.vstack(embeddings)
```

### Qdrant ?游?

```python
# src/vectorization/qdrant_manager.py
from qdrant_client import QdrantClient
from qdrant_client.models import (
    PointStruct, VectorParams, Distance, 
    NamedVector, PayloadSchemaType
)
from typing import Dict, Any, List
import uuid

class QdrantVectorManager:
    "
    Manager for Qdrant vector database operations.
    Handles dual vector storage (original + mask).
    "
    
    def __init__(self, host: str = localhost, port: int = 6333):
        self.client = QdrantClient(host=host, port=port)
        
    def create_collection(self, collection_name: str):
        "
        Create Qdrant collection with dual vector support.
        "
        self.client.recreate_collection(
            collection_name=collection_name,
            vectors_config={
                original: VectorParams(
                    size=512,
                    distance=Distance.COSINE
                ),
                mask: VectorParams(
                    size=512,
                    distance=Distance.COSINE
                )
            }
        )
        
    def insert_dual_vector(self,
                          collection_name: str,
                          original_vector: np.ndarray,
                          mask_vector: np.ndarray,
                          metadata: Dict[str, Any]) -> str:
        "
        Insert dual vectors with metadata.
        
        Args:
            collection_name: Target collection
            original_vector: 512-dim original image vector
            mask_vector: 512-dim mask image vector
            metadata: Additional metadata (file_path, defect_type, etc.)
            
        Returns:
            UUID of inserted point
        "
        point_id = str(uuid.uuid4())
        
        self.client.upsert(
            collection_name=collection_name,
            points=[
                PointStruct(
                    id=point_id,
                    vector={
                        original: original_vector.tolist(),
                        mask: mask_vector.tolist()
                    },
                    payload=metadata
                )
            ]
        )
        
        return point_id
    
    def batch_insert(self,
                    collection_name: str,
                    vectors_data: List[Dict[str, Any]],
                    batch_size: int = 100):
        "
        Batch insert for efficiency.
        
        Args:
            collection_name: Target collection
            vectors_data: List of dicts with keys: original_vector, mask_vector, metadata
            batch_size: Batch size for insertion
        "
        for i in range(0, len(vectors_data), batch_size):
            batch = vectors_data[i:i+batch_size]
            points = [
                PointStruct(
                    id=str(uuid.uuid4()),
                    vector={
                        original: data[original_vector].tolist(),
                        mask: data[mask_vector].tolist()
                    },
                    payload=data[metadata]
                )
                for data in batch
            ]
            self.client.upsert(collection_name=collection_name, points=points)
```

### ????蝔??

```python
# src/vectorization/vectorization_pipeline.py
from pathlib import Path
from PIL import Image
from typing import Optional
import logging

class VectorizationPipeline:
    "
    End-to-end vectorization pipeline integrating CLIP and Qdrant.
    "
    
    def __init__(self, 
                 qdrant_host: str = localhost,
                 qdrant_port: int = 6333,
                 collection_name: str = aoi_defects,
                 device: str = cuda):
        self.vectorizer = CLIPVectorizer(device=device)
        self.qdrant = QdrantVectorManager(host=qdrant_host, port=qdrant_port)
        self.collection_name = collection_name
        self.logger = logging.getLogger(__name__)
        
    def initialize_collection(self):
        "Initialize Qdrant collection if not exists."
        self.qdrant.create_collection(self.collection_name)
        self.logger.info(fCollection
